### PR TITLE
fota - various fixes with the goal to set a proper result once fota aborts

### PR DESCRIPTION
**Pull request by chmielewskiandreas**

### CODE REVIEW

File: subsys/net/lib/download_client/src/download_client.c
Comment (SeppoTakalo) [inline subsys/net/lib/download_client/src/download_client.c:465]: This change is correct, as I don't see `coap_initiate_retransmission()` setting any errno values.

However, I think the correct return code would be `ETIMEDOUT` as we cannot really claim that someone aborted the connection.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:351]: Missing `/* FALLTHRU */`

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:398]: This might be printed even if `configure_full_modem_update()` failed.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]
Thanks for the contribution. I left few remarks.

File: include/net/fota_download.h
Comment (chmielewskiandreas) [reply include/net/fota_download.h:61]: connection lost means that during download there were some connection issues (no network, network issues causing too many retries, socket issues (dl server does not respond). download failed is more caused by configuration/internal issues (invalid packages received etc). 

For an administrator this information is quite useful for troubleshooting. I hope you see my point :)

Comment (SeppoTakalo) [reply include/net/fota_download.h:61]: Well, not really.
You seem to be trying to create distinction between download failures and network failures. In reality, this might not be possible. Download libraries only use Socket API, and they are not following the state of any network interfaces.

I would personally prefer to keep this API generic and simple. So if the outcome of the `CONNECTION_LOST` and `DOWNLOAD_FAILED` is exactly the same. You react exactly same, then those error codes should be the same.

Comment (chmielewskiandreas) [reply include/net/fota_download.h:61]: Take a look at this routine you find in zephyr/subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c. I understand you point to make the interface simple. The code you see below illustrates why I still would like to distinguish between `CONNECTION_LOST` and `DOWNLOAD_FAILED`. For the download client this is valid for the function `coap_initiate_retransmission` you find in `sdk-nrf/subsys/net/lib/download_client/src/coap.c` that limits the number of retry messages. For me this is a big difference between if we stop the download manually or just an error occurs during the download. If e.g. some retry count exceeds the user can actively fix this issue by improving the connection (moving the device to a different location) or simply incrementing the retry count. With my background of supplying thousands of device with a fw update for many years, it was always important for me to know as much as I can why an update failed. Sending just a `DOWNLOAD_FAILED` is somewhat of generic and does not show the reason for a failing update.

```
static void do_transmit_timeout_cb(struct lwm2m_message *msg)
{
    int ret;

    if (firmware_retry < PACKET_TRANSFER_RETRY_MAX) {
        /* retry block */
        LOG_WRN("TIMEOUT - Sending a retry packet!");

        ret = transfer_request(&firmware_block_ctx,
                       msg->token, msg->tkl,
                       do_firmware_transfer_reply_cb);
        if (ret < 0) {
            /* abort retries / transfer */
            set_update_result_from_error(ret);
            firmware_retry = PACKET_TRANSFER_RETRY_MAX;
            lwm2m_engine_context_close(&firmware_ctx);
            return;
        }

        firmware_retry++;
    } else {
        LOG_ERR("TIMEOUT - Too many retry packet attempts! "
            "Aborting firmware download.");
        lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
        lwm2m_engine_context_close(&firmware_ctx);
    }
}
```

**Review by VeijoPesonen — APPROVED** [review]
Just some nitpicking from my side, otherwise looks good to me.

File: include/net/fota_download.h
Comment (SeppoTakalo) [reply include/net/fota_download.h:61]: OK.
I think that is a fair point.
So, again, what is the distinction? Do we mark timeouts as `CONNECTION_LOST` and server returned errors, like file not found, or access denied/socket closed etc. as `DOWNLOAD_FAILED`. Do we then also need to make the distinction that for FAILED case the retry might not help, and for CONNECTION_LOST, you should try rettrying.


Lets wait comments from other reviewers.

I would like for example @lemrey @tokangas to reply. At least you guys have been working with fota_download library.

File: subsys/net/lib/fota_download/src/fota_download.c
Comment (SeppoTakalo) [inline subsys/net/lib/fota_download/src/fota_download.c:247]: Should we make here the distinction that TIMEOUT is a CONNECTION_LOST and the if peer is actively closing the connection (ECONNRESET) then it is DOWNLOAD_FAILED?

File: include/net/fota_download.h
Comment (tokangas) [reply include/net/fota_download.h:61]: I understand the point, but I'm not sure I understand the reasoning for the actual changes in the PR.

The new error cause `FOTA_DOWNLOAD_ERROR_CAUSE_CONNECTION_LOST` is used only in one case, if Download client returns ECONNRESET or ETIMEDOUT. What about for example the connection establishment?

Looking at where `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED` is currently used, it's either for errors from the Download client or the DFU target library. Are the errors from the DFU target library related to the download? Maybe this is a wrong cause to use for those? Maybe at least part of these could be categorized differently.

Most (if not all) errors from the Download client can be categorized as download failures and can be transient in nature, I think. For example, DNS query, TCP connection establishment, TLS negotiation can all fail due to a transient connectivity, network or server issue.

To put it this the other way around, which one of the errors currently mapped to `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED` is such that the application should for sure never retry? I'm not familiar enough with the DFU target library to say how each of the errors from the library should be handled.

Comment (chmielewskiandreas) [reply include/net/fota_download.h:61]: Thanks for your input. I see that its quite difficult to distinguish between CONNECTION_LOST and DOWNLOAD_FAILED. I think I drop changes related to that and put this changes on my ToDo-List. Once I have a clear idea of what has to be done and how, I will open another PR for that. Thanks again :)

**Comment by chmielewskiandreas** [comment]
Hello, I want to ask for updates on this PR.

**Review by lemrey — COMMENTED** [review]
I think one distinction is whether the download could be started or not.
If it could be started, then we incurred in a download issue, and the operation can be retried.
If the download did not start, then it cannot be retried.

Looking at the transitions between state 0, `RESULT_DEFAULT` and state 1, `RESULT_DOWNLOADING`, these are network-related errors that we need to be able to distinguish:
- Update result `4` -> Connection lost during download
- Update result `6` -> Unsupported package type
- Update result `7` -> Invalid URI
- Update result `9` -> Unsupported protocol

If the download can be started, i.e. the package type is supported, the protocol is supported, and the URI is valid, then we can only incur in "Connection lost during download", for which there is an error already (`FOTA_DOWNLOAD_ERROR_CAUSE_CONNECTION_LOST`).
So, error `4` is the only transient one.

To set `RESULT_UPDATE_FAILED` (`5/0/5 = 8`), the state machine must be in state 3, `UPDATING`, which means, the download must have completed. Then it would be incorrect to set result on `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED` since that would indicate the opposite.

**Comment by andreas-chmielewski-gcx** [comment]
@lemrey : Thanks for your comment. Indeed you are right about the state transitions. Thx for your valuable input here. I did just a minor change to set CONNECTION_LOST result on DOWNLOAD_FAILED event. I see some improvements like proper error handling for invalid uri etc. I would not like to touch this for now. This can be done later on in an other patch.

**Comment by chmielewskiandreas** [comment]
Can you please tell me why Jenkins integration fails here and what is still needed to get this PR merged. Thx :)

**Comment by lemrey** [comment]
Hello @chmielewskiandreas. The failures that I see in the last CI run seem unrelated to your PR. It seems CI was run end of June last, I recommend to give this a rebase so we can get a more recent result and see if anything is still red.

**Comment by lemrey** [comment]
There seems to be an error in the documentation stage:
```
Sphinx parallel build error:
sphinx.errors.ExtensionError: Handler <function process_tables at 0x7f775995b550> for event 'source-read' threw an exception (exception: __init__() takes 2 positional arguments but 3 were given)
FAILED: CMakeFiles/matter-all /home/runner/work/sdk-nrf/sdk-nrf/ncs/nrf/doc/_build/CMakeFiles/matter-all 
```
It does not seem to be related to this PR. I'll ask around :)

**Comment by lemrey** [comment]
Hello @chmielewskiandreas, there was an issue with CI builds in NCS which was fixed by this PR: https://github.com/nrfconnect/sdk-nrf/pull/8237
I felt bad asking you to rebase twice in a day, so I have rebased your branch myself this time. I hope that's okay.

**Comment by chmielewskiandreas** [comment]
> Hello @chmielewskiandreas, there was an issue with CI builds in NCS which was fixed by this PR: #8237 I felt bad asking you to rebase twice in a day, so I have rebased your branch myself this time. I hope that's okay.

of course :) Thanks a lot!

**Review by lemrey — APPROVED** [review]
And it's green. This all looks okay to me, but I wonder about the addition of `FOTA_DOWNLOAD_ERROR_CAUSE_CONNECTION_LOST` which is handled in the same way as `FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED`. I'll leave that to you and my colleagues who participated in the review (who are on holiday at the moment, but should be back by August). I don't work much with these libraries these days :)

You'll also need an entry in the changelog, which needs tech-writer approval (also back in August).

**Review by SeppoTakalo — APPROVED** [review]

File: include/net/fota_download.h
Comment (rlubos) [inline include/net/fota_download.h:61]: @chmielewskiandreas Wasn't this new even supposed to be removed, according to previous discussion? It doesn't seem to be set anywhere.

**Review by rlubos — COMMENTED** [review]
Just one concern, looks good otherwise. Please add a changelog entry (https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/releases/release-notes-changelog.rst).

File: include/net/fota_download.h
Comment (chmielewskiandreas) [reply include/net/fota_download.h:61]: Right, I can remove this to avoid dead code

**Review by SeppoTakalo — APPROVED** [review]

**Comment by chmielewskiandreas** [comment]
> Just one concern, looks good otherwise. Please add a changelog entry (https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/releases/release-notes-changelog.rst).

i did so :)

File: doc/nrf/releases/release-notes-changelog.rst
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:485]: back to the LwM2M server?
```suggestion
      * Setting of the FOTA update result.
      * Reporting of the FOTA update result back to the LwM2M server.
```

**Review by SeppoTakalo — APPROVED** [review]

**Review by umapraseeda — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**
